src/orquestra/opt/optimizers/tensor_train_optimizer.py

```diff
@@ -0,0 +1,264 @@
+        return cost
+
+    def gather_best_candidates(self) -> List[dict]:
+        """
+        Creates a list of the best candidates found so far.
+        """
+        return [c for c in self.candidates if "parameters" in c.keys()]
+
+
+class TensorTrainOptimizer(Optimizer):
+    def __init__(
+        self,
+        n_grid_points: Union[int, Sequence[int]],
+        n_evaluations: int,
+        bounds: Bounds,
+        maximum_tensor_train_rank: int = 4,
+        n_rounds: int = 1,
+        maximum_number_of_candidates: int = 5,
+        recorder: RecorderFactory = _recorder,
+    ):
+        """
+        Constructor of a TensorTrainOptimizer. This optimizer uses a tensor-train
+        representation of the cost function to find the minimum of the cost function.
+        It uses the `ttopt` library to do so. This optimizer finds the minimum of the
+        cost function at points given in a grid. This class extends this concept by
+        being able to pick the best candidate solutions in the grid, and then one
+        simply iterates over those candidates on finer grids centered around each
+        candidate to replace each candidate with a better one found in the finer grid.
+
+        Args:
+            n_grid_points: Number of grid points to use in each dimension.
+            n_evaluations: Cost function evaluations budget.
+            bounds: Lower and upper bounds for each parameter.
+            maximum_tensor_train_rank: Maximum bond dimension for the tensor train.
+            n_rounds: Number of optimisation rounds. If 1, then only the initial grid
+                is used. If 2 or more, then the best candidates found in the initial
+                grid are used to create finer grids. Each of these finer grids will
+                get finer and finer depending on the requested number of rounds.
+            maximum_number_of_candidates: Number of candidates to keep track of.
+            recorder Recorder factory for keeping history of calls to the objective
+                function.
+        """
+        super().__init__(recorder=recorder)
+        self.n_grid_points = n_grid_points
+        self.n_evaluations = n_evaluations
+        self.bounds = _get_bounds_like_array(bounds)
+        self.maximum_tensor_train_rank = maximum_tensor_train_rank
+        self.n_rounds = n_rounds
+        self.maximum_number_of_candidates = maximum_number_of_candidates
+        self.evaluations_per_candidate = _get_n_evaluations_per_candidate(
+            n_evaluations, n_rounds, maximum_number_of_candidates
+        )
+
+    def _preprocess_cost_function(
+        self, cost_function: CostFunction
+    ) -> _CostFunctionWithBestCandidates:
+        """
+        Wraps the cost function in a CostFunctionWithBestCandidates.
+        """
+        return _CostFunctionWithBestCandidates(
+            cost_function, self.maximum_number_of_candidates, self.bounds
+        )
```

Comment:
Where is this function being used? It's private, right? So it shouldn't be used somewhere in this PR?